services/filesstore/s3store.go

```diff
@@ -73,20 +78,45 @@ func (b *S3FileBackend) TestConnection() *model.AppError {
 		return model.NewAppError("TestFileConnection", "api.file.test_connection.s3.connection.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
-	exists, err := s3Clnt.BucketExists(context.Background(), b.bucket)
-	if err != nil {
-		return model.NewAppError("TestFileConnection", "api.file.test_connection.s3.bucket_exists.app_error", nil, err.Error(), http.StatusInternalServerError)
+	var exists bool
+	// If a path prefix is present, we attempt to test the bucket by listing objects under the path
+	// and breaking after the first non-nil error. This is because the BucketExists call is only at a bucket level
+	// and sometimes the user might only be allowed access to the specified path prefix.
+	if b.pathPrefix != "" {
+		// In the case where the bucket exists, but the prefix doesn't, and the user has permissions for the entire bucket,
+		// this call cannot differentiate between bucket doesn't exist and prefix doesn't exist.
+		// In that case, it will just redundantly try to create the bucket. It's possible to avoid this with yet another call,
+		// but it's the cost of another network call either ways.
+		objInfoChan := s3Clnt.ListObjects(context.Background(), b.bucket, s3.ListObjectsOptions{Prefix: b.pathPrefix})
+		for obj := range objInfoChan {
```

Comment:
Do we actually need to range? Since we care only about the first object returned we could do a simple receive.

Reply:
Yeah much better. I also improved the logic a bit and fixed the case of prefix doesn't exist. It was a misunderstanding of mine. The range loop was actually exiting without even going inside, so `exists` would remain false. Inverted the boolean value to improve it.